### ISSUE TITLE
Account name changing not working

### ISSUE THREAD

**Issue by wojtekszpunar** [aws-ia/terraform-aws-control_tower_account_factory#43]
I have deployed AFT recently and created a new account.
Unfortunately, I made a typo in AccountName and account_customizations_name, which resulted in the account being created under the wrong name and pointing to a directory which didn't exist.
Strangely enough, Code Pipeline didn't complain about not existing folder and happily reported that everything went OK even though it didn't. 

So I thought easy fix, just update account request and then customizations again. But I was surprised that after merging change to main branch even though terraform updated data in DynamoDB table nothing else happened. 
I mean, no action was triggered on Control Tower to change name and name is still incorrect. 
Another weird part is that when I ran Step Functions to made customizations, instead of looking in aft_request dynamo table, it's parsing aft_request_metadata table. 
I have checked values in that table and those are still invalid and contain old values. Because of that step functions or what ever happens next do not trigger action.

This means some parts of that pipeline didn't work, and some actions weren't triggered

**Comment by stumins** [comment]
Hi, changing the name of an existing account after it has been provisioned is currently not supported by AFT.

https://docs.aws.amazon.com/controltower/latest/userguide/taf-account-provisioning.html#aft-update-account

**Comment by wojtekszpunar** [comment]
Hi,
Thanks for this. 
I'm afraid this is not the only problem described in the above post. 
Yesterday after having longer conversation with AWS Support we have tried to make a workaround to mitigate this issue. Unfortunately, it failed too.

So what I tried to do is to update account name manually since it's just metadata in Control Tower.
Then I have updated manually `aft-request-metadata` DynamoDB table with correct account name and correct `account_customizations_name`.
Next I made a dummy commit to account-customisation repo just to make sure there is some change, thinking it might be related.
Then went ahead and ran `aft-invoke-customizations` step function giving payload:
```
{
  "include": [
    {
      "type": "accounts",
      "target_value": [
        "CORRECT-ACCOUNT-NAME"
      ]
    }
  ]
}
```
This should trigger Account Customization pipeline but it didn't.
Flow on step funtions: Identify Targets -> invoke Provisioning Framework -> Get pipeline Executions -> Pending pipeline executions -> notify success

Basically, step function decided not to invoke pipeline at all.

I have tried to invoke it manually, but unfortunately this is not possible from console.
Pipeline is in Running state. If you try to stop it, the console won't let you because you do not have any pipelines to choose from.

So basically, the entire workflow to apply customizations is not being triggered at all.

How and what does update `aft-request-metadata` table?

**Comment by stumins** [comment]
Thank you for the detailed feedback. First, I want to note that AFT does not support directly changing DynamoDB entries.

However, I’ll try to describe the flow of requests through AFT in concert with the behavior you are seeing.

When a new commit is pushed to the `aft-account-request` repository, a pipeline creates or updates an entry in the `aft-request` DynamoDB table based on the inputs to the `aft` module.

Changes to this table create events that are processed by the `aft-account-request-action-trigger` Lambda function to determine if this request is to create a new account or to update an existing account. For account update requests, this function also determines if the modification requires any Control Tower (CT) changes (e.g. if any of the [control_tower_parameters](https://docs.aws.amazon.com/controltower/latest/userguide/taf-account-provisioning.html#aft-provision-account) in the request were changed).

**Changing an account name**

If Control Tower changes are required, `aft-account-request-action-trigger` places a message in the `aft-account-request` SQS queue, which is processed by the `aft-account-request-processor` Lambda.

`AccountName` is a CT parameter, so when you updated the account name in your request and re-pushed, the request took this path - as an update requiring CT changes.

Because [AFT does not support updating any `control_tower_parameters`](https://docs.aws.amazon.com/controltower/latest/userguide/taf-account-provisioning.html#aft-update-account) other than the `ManagedOrganizationalUnit` property, this change was considered an [invalid modify request](https://github.com/aws-ia/terraform-aws-control_tower_account_factory/blob/main/modules/aft-account-request-framework/lambda/aft-account-request-processor/aft_account_request_processor.py#L59) by the request processor.

I suspect you will find a message in the `aft-account-request-dlq` queue related to this event.

**Re-invoking AFT customizations**

When you invoke the `aft-invoke-customizations` step function, a Lambda function reads the account metadata table to identify all active AFT accounts. The metadata table is updated by the `aft-account-provisioning-framework` Step Function, which only runs after account requests have been fully validated.

Because your account name update request was considered invalid, the metadata table was never updated. When `aft-account-provisioning-framework` ran, it did not find any accounts with “CORRECT-ACCOUNT-NAME”, and as such did not run any customization pipelines.

**Comment by wojtekszpunar** [comment]
Thanks for detailed response. It would be great it that would be described in documentation. 
I mean, how does workflow work (probably README in each of repository?) so it would actually help people with more advanced usages of AFT.

BTW, you have mentioned that I have updated AccountName property and message should be discarded. 
But I haven't only update this property. I did update another one `account_customizations_name`.

So what happens if more than one property is being changed? From execution, I know that even though I have changed customizations, the message was ignored.

With comment of account ID. I've checked documentation again, and I see there is a small note about Account ID being required(although it's very easily to be missed). Can you guys update the JSON file in documentation to reflect that?
I mean when you read this object you automatically think Account Name is required:
```
    {
      "type": "accounts",
      "target_value": [ "acc1","acc2"]
    }
```
If you say `acc1_ID` that changes quite a bit.

**Comment by stumins** [comment]
I've created a few backlog items to improve our documentation:

- Describe how requests flow through AFT
- Clarify the syntax & examples for invoking the customization step function 

I'm going to close this issue now - please feel free to open additional issues if you run into further problems.

**Comment by wojtekszpunar** [comment]
Wouldn't be nicer and more visible to the public if you would only close issues when they are actually done?

**Comment by thebigcosinus** [comment]
Hi,

Did you find a manual solution to update the account Name ?

Thanks for your help

**Comment by wojtekszpunar** [comment]
> Hi,
> 
> Did you find a manual solution to update the account Name ?
> 
> Thanks for your help

Yes. As explained above you need to log in to AWS Account with Root permission (reset password to Root) and just change account name. The change will be automatically recognized in your SSO and ControlTower.
